File: src/main/java/com/hazza/LeetCodeByJava/SegmentOne/No25ReverseNodesInKGroup.java
package com.hazza.LeetCodeByJava.SegmentOne;

import com.hazza.LeetCodeByJava.datastructure.ListNode;

/**
 * https://leetcode.com/problems/reverse-nodes-in-k-group/#/description
 */
public class No25ReverseNodesInKGroup {
    //solution 1
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k <= 0) return head;
        ListNode curr = head;
        int count = 0;
        for (; curr != null && count != k; ++count) curr = curr.next;   //循环到k+1个结点

        if (count == k) {   //保证k+1个结点存在，来保证存在k个结点可供reverse
            curr = reverseKGroup(curr, k);  //递归
            while (count-- > 0) {   //开始逆置
                ListNode temp = head.next;
                head.next = curr;
                curr = head;
                head = temp;
            }
            head = curr;
        }

        return head;
    }

    //solution 2
    public ListNode reverseKGroup2(ListNode head, int k) {
        if (head == null || k <= 0) return head;
        int n = 0;
        for (ListNode i = head; i != null; n++) i = i.next; //计算出结点总个数

        ListNode first = new ListNode(0);
        first.next = head;
        for(ListNode prev = first, tail = head; n >= k; n -= k) {   //递减，将k个结点分组
            for (int i = 1; i < k; i++) {   //每次将下一个结点插到prev后面
                ListNode next = tail.next.next;
                tail.next.next = prev.next;
                prev.next = tail.next;
                tail.next = next;
            }

            prev = tail;
            tail = tail.next;
        }
        return first.next;
    }
}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No25_Reverse_Nodes_in_k-Group.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/21 20:01
# https://leetcode.com/problems/reverse-nodes-in-k-group/
from src.main.python.com.hazza.LeetCodeByPython.dataStructure.ListNode import ListNode, makeListNode, printListNode


class Solution:
    # solution 1, recursively
    def reverseKGroup1(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None

        cur = head
        cnt = 0
        while cur and cnt < k:
            cnt += 1
            cur = cur.next

        if cnt == k:
            cur = self.reverseKGroup1(cur, k)
            for _ in range(k):
                head.next, head, cur = cur, head.next, head
            head = cur

        return head

    # solution 2, iteratively
    def reverseKGroup2(self, head: ListNode, k: int) -> ListNode:
        res = split = ListNode(0)
        res.next = l = r = head

        while True:
            cnt = 0
            while r and cnt < k:
                cnt += 1
                r = r.next
            if cnt == k:
                pre, cur = r, l
                for _ in range(k):
                    cur.next, cur, pre = pre, cur.next, cur
                split.next, split, l = pre, l, r
            else:
                return res.next


if __name__ == '__main__':
    solution = Solution()

    listnode = makeListNode([1, 2, 3, 4, 5])
    printListNode(solution.reverseKGroup1(listnode, 2))
    print()
    listnode = makeListNode([1, 2, 3, 4, 5])
    printListNode(solution.reverseKGroup1(listnode, 3))

    listnode = makeListNode([1, 2, 3, 4, 5])
    printListNode(solution.reverseKGroup2(listnode, 2))
    print()
    listnode = makeListNode([1, 2, 3, 4, 5])
    printListNode(solution.reverseKGroup2(listnode, 3))
